### PR TITLE
Support log density functions as models

### PR DESCRIPTION
**Pull request by devmotion**
This PR adds implementations of `sample` for generic log density functions. It is assumed that models in the `sample` calls that are not subtypes of `AbstractModel` are log density functions. They are supposed to support the LogDensityProblems interface (this is checked) and wrapped in a `LogDensityModel` automatically, which is then used instead of the log density in the `sample` call.

The motivation is that in downstream PRs to me it seemed a bit inconvenient that users have to wrap their log density functions manually. By adding aupport for automatic wrapping to AbatractMCMC instead of downstream packages, all downstream packages can benefit from it without additional code in downstream packages (they only have to support LogDensityModel).

A potential problem could be method ambiguity issues. However, since AbstractSampler is owned by AbstractMCMC this would only be partly AbstractMCMC's fault and I assume it won't happen in practice (let's wait for the downstream tests).

### CODE REVIEW

File: src/sample.jl [hunk 1]
```diff
@@ -85,6 +120,36 @@ function StatsBase.sample(
     return mcmcsample(rng, model, sampler, parallel, N, nchains; kwargs...)
 end
 
+# Fallback: Wrap log density function in a model
+"""
+    sample(
+        rng::Random.AbstractRNG=Random.default_rng(),,
```
Comment (torfjelde) [inline src/sample.jl:126]:
```suggestion
        rng::Random.AbstractRNG=Random.default_rng(),
```

**Review by torfjelde — APPROVED** [review]
LGTM if downstream tests are passing! I only have one change relating to a typo in the docstring, but otherwise es bueno.

**Comment by devmotion** [comment]
I wonder if we could/should add some tests in AbstractMCMC, in addition to the existing downstream tests.

**Review by yebai — APPROVED** [review]

**Comment by torfjelde** [comment]
Though I agree tests would be ideal, if you're busy atm I'm happy to
merge as is given that downstream tests pass.

**Review by cpfiffer — APPROVED** [review]
Looks like a net positive to me, and for relatively low cost. I say we merge it in!

**Comment by torfjelde** [comment]
Oh actually one question @devmotion : do we need `LogDensityModel` after this PR?

**Comment by devmotion** [comment]
I would say, yes, it's still useful since it allows downstream packages to clearly state the `AbstractModel`-type(s) that they want to support with their sampler(s). Additionally, a dedicated model type for log densities makes it easier to define conversion from/to other model types.

**Comment by devmotion** [comment]
I'll see if I find time to add some tests tomorrow. Otherwise I'll merge on Monday, if there are no objections.

**Comment by devmotion** [comment]
I added tests but let's merge https://github.com/TuringLang/AbstractMCMC.jl/pull/114 first (I noticed CompatHelper was disabled due to inactivity).

**Review by torfjelde — APPROVED** [review]
LGTM! Happy to merge as soon as tests pass. Awesome stuff @devmotion :)

**Pull request merged**